### ISSUE TITLE
thread 'main' panicked at 'Wasn't able to create directory::

### ISSUE THREAD

**Issue by jonassmedegaard** [xylous/settle#4]
No matter which subcommand I try to execute, even `settle compl bash`, results in a panick:
```
thread 'main' panicked at 'Wasn't able to create directory:: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', src/io.rs:68:35
```

Full disclosure: I have patched to source to link with an earlier clap v4.1.13, because that's what is a vailable in Debian unstable so far (this is settle packaged officially for Debian).  I have not experienced problems with such hack before, but regardless it is of course not for you to solve if you suspect that I have broken the code by doing that.

**Comment by xylous** [comment]
This is most likely an issue with the configuration, which is _always_ loaded when the program is ran. Regardless, first I'm going to try patching the I/O panics and printing more convenient error messages instead.

**Comment by tptcu** [comment]
This is certainly a configuration error, since it _always_ gets loaded when the program is ran. Needless to say, I added a fix that adds the path to the panic messages, because apparently I've overlooked that.

I've never encountered this error; it's likely that the `$XDG_CONFIG_HOME` (if set) or `$HOME/.config` (if the former is not set) directory permissions don't allow writing and, by extension, creating directories.

**Comment by jonassmedegaard** [comment]
The source code line of the panick message hints that it is calling an mkdir function.

It seems wrong to me that computing shell completion code has the side-effect of messing with the filesystem at all.

**Comment by jonassmedegaard** [comment]
Also, I have no configuration file yet - perhaps you haven't tested that scenario recently (e.g. if you test on same system as you actively use the tool).

**Comment by tptcu** [comment]
As I said, the configuration gets loaded regardless of the command that is being ran, and that's because of [this line in the `main()` function](https://github.com/xylous/settle/blob/master/src/main.rs#L39) and, to some extent, due to the way that the commands are being processed.

As with regards to the configuration file itself, the code tries creating one if it doesn't exist - but in order to create the configuration file, the configuration directory must be made first. Without write permissions, it can't create it. This isn't a problem that the code itself can fix, unfortunately. A possible fix would be loading the default configuration parameters for when something fails along the way. However, that locks `$HOME/zettelkasten` as the Zettelkasten directory, and it doesn't allow using templates. So, it wouldn't be a viable fix.

Unfortunately, I don't really have the means, or the time, to test this program on multiple systems, but I'm going to try and solve any issues that are reported to me to the best of my abilities.

**Comment by jonassmedegaard** [comment]
I did not mean to imply that you should do more testing - I only meant to hint at what you might have missed in your current testing efforts.

It sounds odd that the zettelkasten configuration option affects shell completion code.  Does that mean that after changing zettelkasten path the user needs to regenerate shell completion?!?

I think it would be sensible for the `compl` subcommand to *not* auto-create a default external configuration file, but consult only internal defaults, if at all possible.

**Comment by tptcu** [comment]
Internally, configuration options don't affect shell completion code since they aren't passed to the function that handles it. The fact that the configuration is loaded before the commands are ran is a mere convenience, it makes it easier to not have code repeat. I do agree with this proposal; implementing it should be rather trivial.

**Comment by tptcu** [comment]
This should be it. I also found a bug whereby `compl` was not using the correct parameter name and fixed it (in another commit).

I'll want to implement a few more bugfixes and features that have been on my backlog before releasing the next version, though.

**Comment by tptcu** [comment]
Alright, I released v0.39.10. I hope it works now?

**Comment by jonassmedegaard** [comment]
Improvements - and thanks to now telling the path it failed to create, it is now revealed that the underlying problem that `$XDG_CONFIG_HOME` is resolved at build time - it should be probed at runtime instead.

My build environment involves changing $HOME, and the newly released settle now fails like this:
```
$ settle sync -g
thread 'main' panicked at 'Can't create directory '/build/settle-0.39.10/debian/.debhelper/generated/_source/home/.config/settle'', src/io.rs:60:54
```

My current home is /home/jonas

I suggest to use crate `directories` to handle this :-)

**Comment by tptcu** [comment]
I did actually realise this was a problem right after publishing the version, since I wanted to make it so that you could control the configuration file's location via an environment variable (`$SETTLE_CONFIG`), [but those changes](https://github.com/xylous/settle/commit/05bb46fa84c1443fa2a796f30950cad69a1c048c) were not ammendable to the already published version. I published the new version to crates.io right about now with these fixes.

**Comment by jonassmedegaard** [comment]
`v0.39.11` (and possibly earlier as well) creates config file `/home/jonas/.config/settle/settle.yaml` but it contains this:
```yaml
---
zettelkasten: /build/settle-0.39.11/debian/.debhelper/generated/_source/home/zettelkasten
template: ""
```
So it seems settle still hardcodes $HOME during build :-(

**Comment by jonassmedegaard** [comment]
Looks like you missed this:
```patch
--- a/src/config.rs
+++ b/src/config.rs
@@ -17,7 +17,7 @@
 {
     fn default() -> ConfigOptions
     {
-        ConfigOptions { zettelkasten: format!("{}/zettelkasten", env!("HOME")),
+        ConfigOptions { zettelkasten: format!("{}/zettelkasten", env::var("HOME").unwrap()),
                         template: String::from("") }
     }
 }
```

**Comment by xylous** [comment]
I did catch it [in the commit that came right after](https://github.com/xylous/settle/commit/6ef9c150e6553f1ff3d48e741928f0cf5106e5fe), I apologise for not attaching both of them.

**Comment by tptcu** [comment]
As with regards to it hardcoding it in the configuration file itself, yes, it seems that I completely overlooked it. I'll fix it and publish it on the next version, as it doesn't seem _that_ urgent, and I'm working on another feature which involves visualizing the Zettelkasten in a web browser - not too different from [Obsidian MD's graph view](https://help.obsidian.md/Plugins/Graph+view)

**Comment by jonassmedegaard** [comment]
Please look again - I really don't think you already caught it - what you link to above changes line 86 onwards, where the patch I wrote above affects line 20.

**Comment by tptcu** [comment]
I apologise for that, I deleted the comment.

**Comment by jonassmedegaard** [comment]
Regarding web visualization - you might find inspiration in [Neuron](https://neuron.zettel.page/) and its successor [Emanote](https://emanote.srid.ca/) as well.

**Comment by jonassmedegaard** [comment]
I've released v0.39.11 to Debian now, with my patch applied - works for me.

**Comment by tptcu** [comment]
Awesome! Thanks for all the feedback - I really appreciate it, because I would have had zero chances of finding all of this on my own.
